src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -291,50 +317,88 @@ void registerNext() {
+      LXListenableParameter currentParentParameter = null;
+      for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
+        if (parameter == null) {
+          this.knobs[i] = null;
+          sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
+          continue;
         }
 
-        int i = 0;
-        boolean isEnabled = false;
-        if (this.device != null) {
-          if (this.effect != null) {
-            isEnabled = this.effect.isEnabled();
-          } else if (this.pattern != null) {
-            isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
-          }
-          for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
-            if (i >= this.knobs.length) {
-              break;
-            }
-            this.knobs[i] = parameter;
-            if (parameter != null) {
-              parameter.addListener(this);
-              sendControlChange(0, DEVICE_KNOB_STYLE + i, parameter.getPolarity() == LXParameter.Polarity.BIPOLAR ? LED_STYLE_BIPOLAR : LED_STYLE_UNIPOLAR);
-              double normalized = (parameter instanceof CompoundParameter) ?
-                ((CompoundParameter) parameter).getBaseNormalized() :
-                parameter.getNormalized();
-              sendControlChange(0, DEVICE_KNOB + i, (int) (normalized * 127));
-            } else {
-              sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
-            }
-            ++i;
+        if (i >= this.knobs.length) {
+          // Subtle: Even if all the knobs are full, we need to keep adding
+          // listeners for all children of the final knob. So only break
+          // if we're up to a parameter with no parent or a different parent.
+          if (parameter.parentParameter == null || parameter.parentParameter != currentParentParameter) {
+            break;
           }
-          this.device.controlSurfaceSemaphore.increment();
         }
-        sendNoteOn(0, DEVICE_ON_OFF, isEnabled ? LED_ON : LED_OFF);
-        while (i < this.knobs.length) {
-          sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
-          ++i;
+
+        parameter.addListener(this);
+
+        LXListenableParameter knobParam = parameter;
+
+        if (parameter.parentParameter != null) {
+          if (parameter.parentParameter != currentParentParameter) {
+            // This parameter is the first child to a parent.
+            // Attach the parent to the knob, and keep track of it
+            // so we can skip this step for upcoming siblings.
+            currentParentParameter = parameter.parentParameter;
+            knobParam = currentParentParameter;
+          } else {
+            // This parameter is the non-first child to a parent.
+            // The parent's was already attached to a knob when
+            // we processed its oldest sibling.
+            continue;
+          }
         }
+        this.knobs[i] = knobParam;
```

Comment:
Won't this array overflow if `i >= this.knobs.length` but the checks above don't `break` or `continue`?

Reply:
I've refactored since you made this, but at least in the current version, the only way we get past the `if (i >= this.knobs.length)` check is if `parent` is non-null and the same as `currentParent`, but in that case, the `continue` is hit.